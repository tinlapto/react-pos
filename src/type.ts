export type Item = {
  id: number;
  title: string;
  photo: string;
  price: number;
};

export type Topping = {
  id: number;
  name: string;
};

export type Size = {
  id: number;
  name: string;
};

export type ItemDetail = Item & {
  toppings: Topping[];
  sizes: Size[];
};

export type CartItem = {
  id: number;
  title: string;
  price: number;
  toppings: Topping[];
  sizes: Size[];
  selectedToppings: number[];
  selectedSize: number;
  count: number;
};
