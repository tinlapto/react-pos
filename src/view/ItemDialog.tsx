import {
  Divider,
  message,
  Modal,
  Checkbox,
  Row,
  Radio,
  Button,
  Skeleton,
} from "antd";
import { useEffect, useMemo, useState } from "react";
import { CheckboxValueType } from "antd/es/checkbox/Group";
import InputNumber from "../component/InputNumber";
import { useGetItem } from "../api";
import { CartItem } from "../type";
import { queryClient } from "../App";

const CheckboxGroup = Checkbox.Group;
const RadioGroup = Radio.Group;

const ItemDialog = ({
  id,
  onAdd,
  onCancel,
}: {
  id: number | undefined;
  onAdd: (item: CartItem) => void;
  onCancel: () => void;
}) => {
  const [selectedToppings, setSelectedToppings] = useState<CheckboxValueType[]>(
    []
  );
  const [selectedSize, setSelectedSize] = useState<number>();
  const [count, setCount] = useState<number>(1);

  const {
    data: pizza,
    isLoading,
    refetch,
  } = useGetItem(
    { id: id as number },
    {
      enabled: false,
      onError: (err) => {
        message.error(err.message);
      },
      keepPreviousData: false,
    }
  );

  const { toppings, sizes } = useMemo(
    () => ({
      toppings: (pizza?.toppings ?? []).map(({ id, name }) => ({
        label: name,
        value: id,
      })),
      sizes: (pizza?.sizes ?? []).map(({ id, name }) => ({
        label: name,
        value: id,
      })),
    }),
    [pizza]
  );

  useEffect(() => {
    if (!id) return;
    queryClient.removeQueries("get-item");
    refetch();
    setSelectedToppings([]);
    setSelectedSize(undefined);
    setCount(1);
  }, [id, refetch]);

  const handleOnAdd = () => {
    onAdd({
      ...pizza,
      selectedToppings,
      selectedSize: selectedSize,
      count,
    } as CartItem);
  };

  return (
    <Modal
      visible={!!id}
      onOk={handleOnAdd}
      onCancel={() => onCancel()}
      title={isLoading ? "Loading..." : pizza?.title || "Unknown"}
      footer={
        <div style={{ display: "flex", width: "100%" }}>
          <div style={{ flex: 1 }}>
            <InputNumber value={count} onChange={setCount} min={1} max={10} />
          </div>
          <Button style={{ marginLeft: 8 }} key="back" onClick={onCancel}>
            Cancel
          </Button>
          <Button
            key="submit"
            type="primary"
            onClick={handleOnAdd}
            disabled={isLoading}
          >
            Add To Cart
          </Button>
        </div>
      }
    >
      {isLoading ? (
        <Skeleton />
      ) : (
        <>
          <Row>
            <Divider orientation="left" orientationMargin="0">
              Topping
            </Divider>
            <CheckboxGroup
              options={toppings}
              value={selectedToppings}
              onChange={setSelectedToppings}
            />
          </Row>
          <Row>
            <Divider orientation="left" orientationMargin="0">
              Size
            </Divider>
            <RadioGroup
              options={sizes}
              value={selectedSize}
              onChange={(e) => setSelectedSize(e.target.value)}
            />
          </Row>
        </>
      )}
    </Modal>
  );
};

export default ItemDialog;
