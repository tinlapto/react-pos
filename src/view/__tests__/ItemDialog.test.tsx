import { render, screen, fireEvent } from "@testing-library/react";
import { ItemDetail } from "../../type";
import ItemDialog from "../ItemDialog";

const pizza: ItemDetail = {
  id: 1,
  title: "test name",
  price: 10,
  photo: "test photo",
  toppings: [{ id: 1, name: "topping 1" }],
  sizes: [{ id: 1, name: "size 1" }],
};

let mockIsLoading = false;
jest.mock("../../api", () => ({
  useGetItem: () => ({
    data: pizza,
    refetch: () => {},
    isLoading: mockIsLoading,
  }),
}));

describe("ItemDialog", () => {
  beforeAll(() => {
    // @ts-ignore
    delete window.matchMedia;
    window.matchMedia = (query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    });
  });

  test("render add", () => {
    const onAdd = jest.fn();
    const onCancel = jest.fn();
    const { asFragment } = render(
      <ItemDialog id={1} onAdd={onAdd} onCancel={onCancel} />,
      { container: document.body }
    );
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("topping 1"));

    fireEvent.click(screen.getByText("size 1"));

    fireEvent.click(screen.getByText("Add To Cart"));

    expect(onAdd).toBeCalledWith({
      ...pizza,
      selectedToppings: [1],
      selectedSize: 1,
      count: 1,
    });
  });

  test("render cancel", () => {
    const onAdd = jest.fn();
    const onCancel = jest.fn();
    render(<ItemDialog id={1} onAdd={onAdd} onCancel={onCancel} />, {
      container: document.body,
    });

    // fireEvent.click(screen.getByText("Cancel"));
    fireEvent.click(screen.getByLabelText("close"));

    expect(onCancel).toBeCalled();
  });

  test("render loading", () => {
    const onAdd = jest.fn();
    const onCancel = jest.fn();
    mockIsLoading = true;
    const { asFragment } = render(
      <ItemDialog id={1} onAdd={onAdd} onCancel={onCancel} />,
      {
        container: document.body,
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
