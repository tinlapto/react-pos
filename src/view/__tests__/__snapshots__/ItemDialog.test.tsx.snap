// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDialog render add 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ant-modal-root"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        aria-labelledby="rcDialogTitle0"
        class="ant-modal-wrap"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ant-modal"
          role="document"
          style="width: 520px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <button
              aria-label="Close"
              class="ant-modal-close"
              type="button"
            >
              <span
                class="ant-modal-close-x"
              >
                <span
                  aria-label="close"
                  class="anticon anticon-close ant-modal-close-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="close"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="rcDialogTitle0"
              >
                test name
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <div
                class="ant-row"
              >
                <div
                  class="ant-divider ant-divider-horizontal ant-divider-with-text ant-divider-with-text-left ant-divider-no-default-orientation-margin-left"
                  role="separator"
                >
                  <span
                    class="ant-divider-inner-text"
                    style="margin-left: 0px;"
                  >
                    Topping
                  </span>
                </div>
                <div
                  class="ant-checkbox-group"
                >
                  <label
                    class="ant-checkbox-wrapper ant-checkbox-group-item"
                  >
                    <span
                      class="ant-checkbox"
                    >
                      <input
                        class="ant-checkbox-input"
                        type="checkbox"
                        value="1"
                      />
                      <span
                        class="ant-checkbox-inner"
                      />
                    </span>
                    <span>
                      topping 1
                    </span>
                  </label>
                </div>
              </div>
              <div
                class="ant-row"
              >
                <div
                  class="ant-divider ant-divider-horizontal ant-divider-with-text ant-divider-with-text-left ant-divider-no-default-orientation-margin-left"
                  role="separator"
                >
                  <span
                    class="ant-divider-inner-text"
                    style="margin-left: 0px;"
                  >
                    Size
                  </span>
                </div>
                <div
                  class="ant-radio-group ant-radio-group-outline"
                >
                  <label
                    class="ant-radio-wrapper"
                  >
                    <span
                      class="ant-radio"
                    >
                      <input
                        class="ant-radio-input"
                        type="radio"
                        value="1"
                      />
                      <span
                        class="ant-radio-inner"
                      />
                    </span>
                    <span>
                      size 1
                    </span>
                  </label>
                </div>
              </div>
            </div>
            <div
              class="ant-modal-footer"
            >
              <div
                style="display: flex; width: 100%;"
              >
                <div
                  style="flex: 1;"
                >
                  <span
                    class="ant-input-group-wrapper"
                    style="text-align: right;"
                  >
                    <span
                      class="ant-input-wrapper ant-input-group"
                    >
                      <span
                        class="ant-input-group-addon"
                      >
                        <span
                          aria-label="plus"
                          class="anticon anticon-plus"
                          role="img"
                          tabindex="-1"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="plus"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <defs>
                              <style />
                            </defs>
                            <path
                              d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                            />
                            <path
                              d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                            />
                          </svg>
                        </span>
                      </span>
                      <input
                        class="ant-input"
                        type="text"
                        value="1"
                      />
                      <span
                        class="ant-input-group-addon"
                      >
                        <span
                          aria-label="minus"
                          class="anticon anticon-minus"
                          role="img"
                          tabindex="-1"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="minus"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                            />
                          </svg>
                        </span>
                      </span>
                    </span>
                  </span>
                </div>
                <button
                  class="ant-btn ant-btn-default"
                  style="margin-left: 8px;"
                  type="button"
                >
                  <span>
                    Cancel
                  </span>
                </button>
                <button
                  class="ant-btn ant-btn-primary"
                  type="button"
                >
                  <span>
                    Add To Cart
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ItemDialog render loading 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ant-modal-root"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        aria-labelledby="rcDialogTitle2"
        class="ant-modal-wrap"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ant-modal"
          role="document"
          style="width: 520px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <button
              aria-label="Close"
              class="ant-modal-close"
              type="button"
            >
              <span
                class="ant-modal-close-x"
              >
                <span
                  aria-label="close"
                  class="anticon anticon-close ant-modal-close-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="close"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="rcDialogTitle2"
              >
                Loading...
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <div
                class="ant-skeleton"
              >
                <div
                  class="ant-skeleton-content"
                >
                  <h3
                    class="ant-skeleton-title"
                    style="width: 38%;"
                  />
                  <ul
                    class="ant-skeleton-paragraph"
                  >
                    <li />
                    <li />
                    <li
                      style="width: 61%;"
                    />
                  </ul>
                </div>
              </div>
            </div>
            <div
              class="ant-modal-footer"
            >
              <div
                style="display: flex; width: 100%;"
              >
                <div
                  style="flex: 1;"
                >
                  <span
                    class="ant-input-group-wrapper"
                    style="text-align: right;"
                  >
                    <span
                      class="ant-input-wrapper ant-input-group"
                    >
                      <span
                        class="ant-input-group-addon"
                      >
                        <span
                          aria-label="plus"
                          class="anticon anticon-plus"
                          role="img"
                          tabindex="-1"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="plus"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <defs>
                              <style />
                            </defs>
                            <path
                              d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                            />
                            <path
                              d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                            />
                          </svg>
                        </span>
                      </span>
                      <input
                        class="ant-input"
                        type="text"
                        value="1"
                      />
                      <span
                        class="ant-input-group-addon"
                      >
                        <span
                          aria-label="minus"
                          class="anticon anticon-minus"
                          role="img"
                          tabindex="-1"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="minus"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                            />
                          </svg>
                        </span>
                      </span>
                    </span>
                  </span>
                </div>
                <button
                  class="ant-btn ant-btn-default"
                  style="margin-left: 8px;"
                  type="button"
                >
                  <span>
                    Cancel
                  </span>
                </button>
                <button
                  class="ant-btn ant-btn-primary"
                  disabled=""
                  type="button"
                >
                  <span>
                    Add To Cart
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
